File: javascript/203.remove-linked-list-elements.js
/*
 * @lc app=leetcode id=203 lang=javascript
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} val
 * @return {ListNode}
 */
var removeElements = function (head, val) {
  // Solution 1:
  // let newHead = new ListNode();
  // let p = head,
  //   q = newHead;
  // while (p) {
  //   if (p.val !== val) {
  //     q.next = p;
  //     q = q.next;
  //   }
  //   p = p.next;
  // }
  // q.next = p;
  // return newHead.next;

  // Solution 2:
  let p = head,
    prev;
  while (p) {
    if (head.val === val) {
      head = head.next;
      p = head;
    } else {
      if (p.val === val) {
        prev.next = p.next;
      } else {
        prev = p;
      }
      p = p.next;
    }
  }
  return head;
};
// @lc code=end

File: rust/203.remove-linked-list-elements.rs
/*
 * @lc app=leetcode id=203 lang=rust
 *
 * [203] Remove Linked List Elements
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn remove_elements(head: Option<Box<ListNode>>, val: i32) -> Option<Box<ListNode>> {
        let mut result = head.clone();
        let mut current = &mut result;
        // idk why `while let` doesn't work so i checked the solution
        // and modified like this:
        loop {
            match current {
                Some(node) if node.val == val => {
                    *current = node.next.take();
                }
                Some(node) => {
                    current = &mut node.next;
                }
                None => break,
            }
        }
        return result;
    }
}
// @lc code=end
